### DISCUSSION TITLE
Some questions about the code

### DISCUSSION THREAD

**Discussion by Carrotljw** [brianhie/viral-mutation#17] [Q&A]
Hello, does the language model of this paper use transformer? I saw this part in the code, but it seems that I did not see the relevant introduction in the paper.Hope to get your answer, thank you~

**Comment by brianhie** [comment]
Hi @Carrotljw, there is some code that tries to implement a transformer, but the main model that we use is a BiLSTM.

**Reply by Carrotljw** [reply]
Thank you very much for your answer~ I saw the BiLSTM model (BiLSTMLanguageModel) in the language model module, but it seems that it is only used for the LSTM module, not BiLSTM.
`        for _ in range(n_hidden - 1):
            lstm = LSTM(hidden_dim, return_sequences=True)
            x_pre = lstm(x_pre)
            x_post = lstm(x_post)
        lstm = LSTM(hidden_dim)
        x_pre = lstm(x_pre)
        x_post = lstm(x_post)`

**Reply by brianhie** [reply]
The way our BiLSTM is implemented, the same LSTM parameters are applied to the pre and post sequence context. So that's why `lstm` is evaluated on both `x_pre` and `x_post`

**Reply by Carrotljw** [reply]
In each training, one of the amino acids in the sequence will be removed, and then the front part and the back part will be passed into the same LSTM to predict what the removed amino acid is. Could I understand it this way?

**Reply by brianhie** [reply]
Yup, that is right!

**Reply by Carrotljw** [reply]
Thank you for your patience to answer~ Your work understands the evolution of the virus from a new perspective. it’s very meaningful and interesting. I hope I could have the opportunity to discuss it with you in the future. 😊

**Reply by Carrotljw** [reply]
Sorry I have new questions, I don't know how semantic embedding is obtained. I read this part of the code but didn't understand it.
My puzzle: Is it to input the original sequence and the mutated sequence into the trained model and get their respective semantic embeddings?

- If so, these sequences are entered together with mutation sites, but the mutation sites are removed during model training, so how do I enter into the model to get semantic embedding😭
- Could you briefly explain the structure of the code？🤓

**Reply by brianhie** [reply]
Please see this function for computing semantic embeddings: https://github.com/brianhie/viral-mutation/blob/27dcc5a019dec123f78e5b9c7bc89bbcd9eed68d/bin/mutation.py#L183. We use the output of the last hidden layer of the BiLSTM as the embeding: https://github.com/brianhie/viral-mutation/blob/27dcc5a019dec123f78e5b9c7bc89bbcd9eed68d/bin/language_model.py#L71
